File: BOJ/python3/2805.py
'''
나무 자르기
https://www.acmicpc.net/problem/2805
'''
import sys
import time
start = time.time()  # 시작 시간
sys.stdin = open("input.txt", "r")  # input.txt 파일로 입력

N, M = map(int, input().split())
trees = list(map(int, input().split()))

max_height = 0
l, r = 1, max(trees)
while l <= r:
    mid = (l+r) // 2

    # 가져갈 나무 길이 구하기
    total = 0
    for tree in trees:
        if tree > mid:
            total += tree - mid

    # 적어도 M미터의 나무를 가져갈 때, 높이의 최대값
    if total >= M:
        if mid > max_height:
            max_height = mid

    # 이진 탐색을 위한 범위 조정
    if total > M:
        l = mid + 1
    else:
        r = mid - 1

print(max_height)

# 현재시각 - 시작시간 = 실행 시간(sec)
print("--------------------")
print(f"Elapsed Time: {round(time.time() - start, 3)}s")

File: BOJ/nodejs/2805.js
/*
나무 자르기
https://boj.kr/2805
 */
const start = new Date().getTime();

const input = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .split("\n");

const [N, M] = input[0].split(" ").map(Number); // [나무의 수, 가져가려고 하는 나무길이]
const heights = input[1].split(" ").map(Number); // 나무의 높이[]

let lowH = 1;
let highH = Math.max(...heights);
let H;
let ans = 0;
while (lowH <= highH) {
  H = Math.floor((lowH + highH) / 2);

  const sum = heights
    .map((height) => height - H > 0 && height - H)
    .reduce((acc, cur) => acc + cur, 0);
  if (sum >= M) {
    ans = Math.max(ans, H);
  }

  sum > M ? (lowH = H + 1) : (highH = H - 1);
}

console.log(ans);

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
